sopt/src/main/scala/dagr/sopt/cmdline/CommandLineParser.scala

```diff
@@ -0,0 +1,527 @@
+    */
+  private def unknownClpError(classes: Set[ClpClass], clpName: String) : String = {
+    val builder = new StringBuilder
+    val distances: mutable.Map[ClpClass, Integer] = new mutable.HashMap[ClpClass, Integer]
+    var bestDistance: Int = Integer.MAX_VALUE
+    var bestN: Int = 0
+    classes.foreach{ clazz =>
+      val name: String = clazz.getSimpleName
+      if (name == clpName) {
+        throw new CommandLineParserInternalException("Command matches when searching for the unknown: " + clpName)
+      }
+      val distance: Int = if (name.startsWith(clpName) || (MinimumSubstringLength <= clpName.length && name.contains(clpName))) {
+        0
+      }
+      else {
+        levenshteinDistance(clpName, name, 0, 2, 1, 4)
+      }
+      distances.put(clazz, distance)
+      if (distance < bestDistance) {
+        bestDistance = distance
+        bestN = 1
+      }
+      else if(distance == bestDistance) {
+        bestN += 1
+      }
+    }
+    if (0 == bestDistance && bestN == classes.size) {
+      bestDistance = HelpSimilarityFloor + 1
+    }
+    builder.append(s"${unknownCommandErrorMessage(clpName)}\n")
+    if (bestDistance < HelpSimilarityFloor) {
+      builder.append(String.format("Did you mean %s?\n", if (bestN < 2) "this" else "one of these"))
+      builder.append(classes.filter(bestDistance == distances.get(_).get).mkString("\n        "))
+    }
+
+    builder.toString()
+  }
+
+  /** Returns the implementation version string of the given class */
+  protected def version: String = version(getClass)
+
+  /**
+    * Prints the top-level usage of the command line, with a standard pre-amble, version, and list of command line
+    * programs available.
+    *
+    * @param classes the classes corresponding to the command line programs.
+    * @param commandLineName the name of this command line program (ex. dagr).
+    */
+  private[cmdline] def clpListUsage(classes: Set[ClpClass], commandLineName: String, withPreamble: Boolean): String = {
+    val builder = new StringBuilder
+
+    if (withPreamble) {
+      builder.append(s"${standardClpUsagePreamble()}\n")
+      builder.append(wrapString(KRED, s"$version\n", KNRM))
+    }
+
+    builder.append(wrapString(KBLDRED, s"$AvailableClps\n", KNRM))
+    val clpGroupClassToClpGroupInstance: mutable.Map[ClpGroupClass, ClpGroup] = new mutable.HashMap[ClpGroupClass, ClpGroup]
+    val clpsByGroup: java.util.Map[ClpGroup, ListBuffer[ClpClass]] = new java.util.TreeMap[ClpGroup, ListBuffer[ClpClass]]()
+    val clpsToProperty: mutable.Map[ClpClass, clp] = new mutable.HashMap[ClpClass, clp]
+
+    classes.foreach { clazz =>
+      findClpAnnotation(clazz) match {
+        case None => throw new BadAnnotationException(s"The class '${clazz.getSimpleName}' is missing the required @clp annotation.")
+        case Some(clp) =>
+          clpsToProperty.put(clazz, clp)
+          val clpGroup: ClpGroup = clpGroupClassToClpGroupInstance.get(clp.group) match {
```

Comment:
Wouldn't it be simpler to work with class objects for groups instead of instances, up until you need the properties for printing down below?